### PR TITLE
feat: Multi-Purpose Tokens (MPT)

### PR DESCRIPTION
**Pull request by khancode**
## High Level Overview of Change

Adds Multi-Purpose Tokens (MPT) feature.

### Context of Change

[XLS-33 spec](https://github.com/XRPLF/XRPL-Standards/tree/master/XLS-0033d-multi-purpose-tokens)

### Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [ ] Tests (You added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation Updates
- [ ] Release

### Did you update CHANGELOG.md?

- [x] Yes
- [ ] No, this change does not impact library users

## Test Plan

Adds unit/integ tests for new transactions.

### CODE REVIEW

File: tests/integration/transactions/test_mptoken_authorize.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
+# TODO: uncomment test module when rippled Docker image supports MPT feature
+# from tests.integration.integration_test_case import IntegrationTestCase
+# from tests.integration.it_utils import (
+#     fund_wallet_async,
+#     sign_and_reliable_submission_async,
+#     test_async_and_sync,
```
Comment (achowdhry-ripple) [inline tests/integration/transactions/test_mptoken_authorize.py:6]: MPT is now in devnet, so there should be a docker image available now to uncomment these

File: tests/integration/transactions/test_clawback.py [hunk 2]
```diff
@@ -71,3 +79,92 @@ async def test_basic_functionality(self, client):
             client,
         )
         self.assertTrue(response.is_successful())
+
+    # TODO: uncomment test when rippled Docker image supports MPT feature
+    # @test_async_and_sync(globals())
+    # async def test_mptoken(self, client):
+    #     wallet2 = Wallet.create()
```
Comment (achowdhry-ripple) [inline tests/integration/transactions/test_clawback.py:86]: Will also need to pull in the changes from the verified credentials PR that edit the ci docker command for the entrypoint change

File: pyproject.toml [hunk 3]
```diff
@@ -1,6 +1,6 @@
 [tool.poetry]
 name = "xrpl-py"
-version = "3.0.0"
+version = "4.0.0b3"
```
Comment (achowdhry-ripple) [inline pyproject.toml:3]: Should this be a minor version upgrade? Will we need another major version upgrade for each new amendment?

File: xrpl/models/transactions/xchain_claim.py [hunk 4]
```diff
@@ -77,7 +78,17 @@ def _get_errors(self: Self) -> Dict[str, str]:
...
+        amount = self.amount
+        if is_xrp(amount):
+            currency = XRP()
+        elif is_issued_currency(amount):
+            currency = amount.to_currency()  # type: ignore
+        elif is_mpt(amount):
+            currency = amount.mpt_issuance_id  # type: ignore
+        else:
+            errors["amount"] = "currency can't be derived."
+
```
Comment (achowdhry-ripple) [inline xrpl/models/transactions/xchain_claim.py:91]: why are these changes needed in python and not the JS lib?

File: xrpl/models/transactions/xchain_claim.py [hunk 5]
```diff
@@ -77,7 +78,17 @@ def _get_errors(self: Self) -> Dict[str, str]:
...
         if (
             currency != bridge.locking_chain_issue
             and currency != bridge.issuing_chain_issue
```
Comment (achowdhry-ripple) [inline xrpl/models/transactions/xchain_claim.py:94]: Seems like this last error isn't tracked in the JS lib. Is this no longer relevant or should it be added to the JS library as well?

File: xrpl/models/transactions/xchain_claim.py (hunk 4 above)
Comment (khancode) [reply xrpl/models/transactions/xchain_claim.py:91]: They're also in JS - https://github.com/XRPLF/xrpl.js/pull/2661/files#diff-695e6507bbf3a86833c7318c48da6ec25fe12c7a4f8a90a3a0cfbb8c19f8b3abR167

File: xrpl/models/transactions/mptoken_issuance_create.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    Note that this value is optional, and will default to 0 if not supplied.
+    """
+
+    maximum_amount: Optional[str] = None
+    """
+    Specifies the maximum asset amount of this token that should ever be issued.
+    It is a non-negative integer that can store a range of up to 63 bits. If
+    not set, the max amount will default to the largest unsigned 63-bit integer
+    (0x7FFFFFFFFFFFFFFF)
+    """
```
[the remark this answers is not in the document]
Comment (khancode) [reply xrpl/models/transactions/mptoken_issuance_create.py:76]: It's a non-negative integer string.

File: xrpl/models/transactions/mptoken_issuance_create.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+    transfer_fee: Optional[int] = None
+    """
+    Specifies the fee to charged by the issuer for secondary sales of the Token,
+    if such sales are allowed. Valid values for this field are between 0 and 50,000
+    inclusive, allowing transfer rates of between 0.000% and 50.000% in increments of
+    0.001. The field must NOT be present if the `tfMPTCanTransfer` flag is not set.
+    """
```
[the remark this answers is not in the document]
Comment (khancode) [reply xrpl/models/transactions/mptoken_issuance_create.py:84]: Flag checks aren't done elsewhere with models.

File: xrpl/models/base_model.py [hunk 8]
```diff
@@ -59,6 +62,9 @@ def _key_to_json(field: str) -> str:
     Raises:
         XRPLModelException: If the input is invalid
     """
+    if field in EXCLUDED_KEYS:
+        return field
+
```
[the remark this answers is not in the document]
Comment (khancode) [reply xrpl/models/base_model.py:67]: It needs to stay in snake_case for request responses - https://github.com/XRPLF/XRPL-Standards/tree/master/XLS-0033d-multi-purpose-tokens#6311-example

File: pyproject.toml (hunk 3 above)
Comment (achowdhry-ripple) [reply pyproject.toml]: Ok sure, so when we do a general release then it'll go to 3.1.0?

Comment (achowdhry-ripple) [reply pyproject.toml]: Also if we need a quick small release while MPT is in beta, could that then be a patch? so if 3.1.0bX is MPT beta, then 3.1.1 would be a change needed while MPT is in beta? Otherwise if we follow this process for upcoming amendments our library version would get very high next year

File: pyproject.toml [hunk 9]
```diff
@@ -1,6 +1,6 @@
 [tool.poetry]
 name = "xrpl-py"
-version = "3.0.0"
+version = "4.0.0"
```
Comment (achowdhry-ripple) [inline pyproject.toml:3]: should this be a beta version with bX postfix?

File: xrpl/models/transactions/mptoken_issuance_set.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+        init=False,
+    )
+
+    def _get_errors(self: Self) -> Dict[str, str]:
+        errors = super()._get_errors()
+
+        if self.has_flag(MPTokenIssuanceSetFlag.TF_MPT_LOCK) and self.has_flag(
+            MPTokenIssuanceSetFlag.TF_MPT_UNLOCK
+        ):
+            errors["flags"] = "flag conflict"
```
Comment (achowdhry-ripple) [inline xrpl/models/transactions/mptoken_issuance_set.py:77]: nit: could we make the error message a bit more specific?

**Comment by achowdhry-ripple** [comment]
Same as the comment from the js library, am assuming the getBalanceChanges utility here would need to be modified in the same way (or have a comment and raise an issue to track the addition later on)

File: pyproject.toml (hunk 9 above)
Comment (khancode) [reply pyproject.toml:3]: Nope, it should be normal major release since MPT is available in devnet and rippled development is completed.

**Review by achowdhry-ripple — APPROVED** [review]
LGTM

File: xrpl/core/binarycodec/types/uint64.py
Comment (shawnxie999) [inline xrpl/core/binarycodec/types/uint64.py:1]: do we have something similar to verify uint64 parsing? 

for context, i have some tests to serialize/deserialize MPT fields in xrpl.js: https://github.com/XRPLF/xrpl.js/blob/b04efe8c9e2aec01b866316d0c72ec85b27a65c9/packages/ripple-binary-codec/test/uint.test.ts#L182

File: tests/unit/models/requests/test_ledger_entry.py [hunk 11]
```diff
@@ -152,3 +152,36 @@ def test_invalid_price_oracle_object(self):
             LedgerEntry(
                 oracle=Oracle(oracle_document_id=1),
             )
+
+    def test_get_mpt_issuance(self):
+        # oracle_document_id is specified as uint
```
Comment (shawnxie999) [inline tests/unit/models/requests/test_ledger_entry.py:157]: is oracle_document_id typo?

**Review by shawnxie999 — COMMENTED** [review]
I think it looks good. Might be useful to test serialization/deserialization of MPTokenIssuance and MPToken objects

File: xrpl/core/binarycodec/types/uint64.py
Comment (khancode) [reply xrpl/core/binarycodec/types/uint64.py:1]: I just added tests in https://github.com/XRPLF/xrpl-py/pull/732/commits/6eba95515e556178c35b928a30e63b09c296f34d

**Comment by khancode** [comment]
> I think it looks good. Might be useful to test serialization/deserialization of MPTokenIssuance and MPToken objects

xrpl-py doesn't support ledger objects yet so that's why they're excluded for now.

**Review by shawnxie999 — APPROVED** [review]
Looks good
